File: apps/ios/Ghostpass/Theme/Theme.swift
import SwiftUI

/// Le système visuel néon de la suite, partagé avec `ghostbit`, `ghostmon`, `ghostboard`,
/// `ghostauth` et `ghostcal`.
///
/// La structure est commune à tous les produits : une base sombre dérivée de Dracula, des
/// surfaces de verre fumé, et un halo coloré diffusé depuis le haut. Ce qui les distingue
/// est leur teinte néon, celle de leur logo — violet pour ghostbit, lime pour ghostmon,
/// orange pour ghostboard, menthe pour ghostauth, cyan pour ghostcal, jaune pour
/// ghostmail. Pour GhostPass, c'est le **bleu `#2E7DFF`**, seul créneau froid encore libre
/// et déjà celui de la marque.
///
/// Le néon n'est pas dans la couleur mais dans son rayonnement : deux halos superposés,
/// l'un serré et vif, l'autre large et discret. Voir `neon()`.
///
/// Les deux palettes existent, comme sur le web : le sombre est la teinte d'origine, le
/// clair une déclinaison de plein jour. On suit le réglage du système plutôt que d'imposer
/// l'un des deux.
extension Color {
    // ─── Base sombre commune à la suite (Dracula) ───

    /// Fond de page.
    static let gpBase = adaptative(sombre: 0x21222C, clair: 0xF5F8FC)
    /// Cartes et lignes.
    static let gpSurface = adaptative(sombre: 0x282A36, clair: 0xFFFFFF)
    /// Champs et surfaces imbriquées.
    static let gpSurface2 = adaptative(sombre: 0x323445, clair: 0xEEF3F9)
    /// Texte principal.
    static let gpInk = adaptative(sombre: 0xF8F8F2, clair: 0x0F1B2D)
    /// Texte secondaire.
    static let gpMuted = adaptative(sombre: 0x8B9CC8, clair: 0x5B6B82)

    // ─── Teinte de marque : bleu néon ───

    /// Accent plein : fonds de boutons, pastilles. Assez sombre pour porter du blanc.
    static let gpAccent = adaptative(sombre: 0x2E7DFF, clair: 0x1A4FCC)
    /// Accent en texte ou en icône. Sur fond de nuit, le bleu plein manque de clarté :
    /// c'est la teinte haute de la marque qui prend le relais.
    static let gpAccentText = adaptative(sombre: 0x7FB2FF, clair: 0x1A4FCC)
    /// Ce qui se pose sur `gpAccent`.
    static let gpOnAccent = Color.white
    /// La teinte du halo. Fixe : un néon garde sa couleur, c'est ce qui le fait lire comme
    /// une source lumineuse plutôt que comme une ombre portée teintée.
    static let gpNeon = Color(rgb: 0x2E7DFF)

    static let gpBorder = adaptativeAlpha(sombre: (0xFFFFFF, 0.08), clair: (0x0F172A, 0.10))
    static let gpBorderStrong = adaptativeAlpha(sombre: (0xFFFFFF, 0.16), clair: (0x0F172A, 0.16))

    /// Sémantiques, reprises telles quelles de la base Dracula partagée.
    static let gpDanger = adaptative(sombre: 0xFF5555, clair: 0xD11F45)
    static let gpSuccess = adaptative(sombre: 0x50FA7B, clair: 0x15803D)

    /// Une teinte fixe, en hexadécimal. Sert aux couleurs qui ne dépendent pas du thème :
    /// les pastilles d'initiale, dont la palette est partagée avec la web app.
    init(rgb: Int) {
        self.init(UIColor(rgb: rgb, alpha: 1))
    }

    private static func adaptative(sombre: Int, clair: Int) -> Color {
        Color(
            UIColor { traits in
                UIColor(rgb: traits.userInterfaceStyle == .dark ? sombre : clair, alpha: 1)
            })
    }

    private static func adaptativeAlpha(
        sombre: (Int, CGFloat), clair: (Int, CGFloat)
    ) -> Color {
        Color(
            UIColor { traits in
                let (rgb, alpha) = traits.userInterfaceStyle == .dark ? sombre : clair
                return UIColor(rgb: rgb, alpha: alpha)
            })
    }
}

extension UIColor {
    fileprivate convenience init(rgb: Int, alpha: CGFloat) {
        self.init(
            red: CGFloat((rgb >> 16) & 0xFF) / 255,
            green: CGFloat((rgb >> 8) & 0xFF) / 255,
            blue: CGFloat(rgb & 0xFF) / 255,
            alpha: alpha)
    }
}

/// Mesures partagées. Les mêmes que sur le web : contrôles à 8, cartes à 16.
enum GP {
    static let radius: CGFloat = 12
    static let radiusCard: CGFloat = 16
    static let gap: CGFloat = 12
    static let padding: CGFloat = 16
    static let paddingCard: CGFloat = 24
}

/// Le halo néon de la suite : deux rayonnements superposés, l'un serré et vif, l'autre
/// large et diffus.
///
/// C'est la transposition exacte du `--brand-glow` partagé — `drop-shadow(0 0 8px …0.6)`
/// puis `drop-shadow(0 0 20px …0.35)`. Les deux comptent : un seul halo fait une tache
/// molle, tandis que la superposition d'un noyau net et d'une aura étalée est ce qui donne
/// l'impression d'une source de lumière. Le rayon est réduit en thème clair, où un néon
/// sur fond blanc devient une bavure.
struct Neon: ViewModifier {
    /// Multiplicateur d'intensité : 1 pour une marque, moins pour un détail.
    var force: Double = 1

    @Environment(\.colorScheme) private var schema

    func body(content: Content) -> some View {
        let echelle = schema == .dark ? force : force * 0.45
        return
            content
            .shadow(color: Color.gpNeon.opacity(0.60 * echelle), radius: 8 * echelle)
            .shadow(color: Color.gpNeon.opacity(0.35 * echelle), radius: 20 * echelle)
    }
}

extension View {
    /// Fait rayonner un élément dans la teinte de la marque.
    func neon(_ force: Double = 1) -> some View { modifier(Neon(force: force)) }
}

/// Le fond de l'application : nuit profonde et halo néon diffusé depuis le haut — la lueur
/// du fantôme, qui donne sa profondeur à l'ensemble sans rien coûter en lisibilité.
///
/// Le dégradé du fond est celui de la suite : quatre arrêts à 150°, teintés de la couleur
/// de marque, qui empêchent le noir d'être plat.
struct GhostBackground: View {
    @Environment(\.colorScheme) private var schema

    var body: some View {
        fond
            .overlay(alignment: .top) {
                RadialGradient(
                    colors: [Color.gpNeon.opacity(schema == .dark ? 0.26 : 0.12), .clear],
                    center: .top, startRadius: 0, endRadius: 520
                )
                .frame(height: 620)
                .offset(y: -140)
                .blur(radius: 40)
            }
            .ignoresSafeArea()
    }

    @ViewBuilder private var fond: some View {
        if schema == .dark {
            LinearGradient(
                colors: [
                    Color(rgb: 0x080D18), Color(rgb: 0x0A1220),
                    Color(rgb: 0x080F1A), Color(rgb: 0x090A10),
                ],
                startPoint: .top, endPoint: .bottom)
        } else {
            Color.gpBase
        }
    }
}

/// Carte de verre fumé : la surface translucide, bordée d'un filet clair.
struct GlassCard: ViewModifier {
    var padding: CGFloat = GP.paddingCard

    func body(content: Content) -> some View {
        content
            .padding(padding)
            .background(Color.gpSurface.opacity(0.65), in: rect)
            .overlay(rect.strokeBorder(Color.gpBorder, lineWidth: 1))
            .background(.ultraThinMaterial, in: rect)
    }

    private var rect: RoundedRectangle { RoundedRectangle(cornerRadius: GP.radiusCard) }
}

extension View {
    func glassCard(padding: CGFloat = GP.paddingCard) -> some View {
        modifier(GlassCard(padding: padding))
    }
}

/// Action principale : un bloc d'accent plein, pleine largeur.
/// Bouton principal. Comme le secondaire, il honore le rôle destructif : un bouton qui
/// détruit ne doit pas ressembler à celui qui valide.
struct PrimaryButtonStyle: ButtonStyle {
    var enabled = true

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(.system(.body, weight: .semibold))
            .foregroundStyle(Color.gpOnAccent)
            .frame(maxWidth: .infinity)
            .padding(.vertical, 14)
            .background(
                (configuration.role == .destructive ? Color.gpDanger : Color.gpAccent)
                    .opacity(enabled ? (configuration.isPressed ? 0.8 : 1) : 0.35),
                in: RoundedRectangle(cornerRadius: GP.radius)
            )
            // Seule l'action disponible rayonne. Faire luire un bouton inerte reviendrait
            // à appeler l'œil vers ce sur quoi on ne peut pas appuyer.
            .neon(enabled ? (configuration.isPressed ? 0.5 : 0.85) : 0)
            .contentShape(RoundedRectangle(cornerRadius: GP.radius))
    }
}

/// Action secondaire : le même bloc, mais creusé plutôt que plein.
struct SecondaryButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(.system(.subheadline, weight: .medium))
            // Le rôle décide de la couleur.
            //
            // Un style personnalisé **écrase** le rendu que SwiftUI donne au rôle
            // destructif : « Supprimer ce groupe » portait bien `role: .destructive` et
            // s'affichait en bleu, comme n'importe quelle action ordinaire. Le défaut
            // paraît juste à la lecture du code et faux à l'écran — c'est celui qu'on ne
            // trouve qu'en regardant.
            //
            // Le lire ici plutôt qu'à chaque appel corrige la classe entière, y compris
            // les boutons qui n'existent pas encore.
            .foregroundStyle(
                configuration.role == .destructive ? Color.gpDanger : Color.gpAccentText
            )
            .frame(maxWidth: .infinity)
            .padding(.vertical, 12)
            .background(
                Color.gpSurface2.opacity(configuration.isPressed ? 0.6 : 1),
                in: RoundedRectangle(cornerRadius: GP.radius)
            )
            .overlay(
                RoundedRectangle(cornerRadius: GP.radius)
                    .strokeBorder(Color.gpBorder, lineWidth: 1))
    }
}

/// Champ de saisie : surface creusée, bordure discrète, texte au repos lisible.
struct GhostFieldStyle: ViewModifier {
    func body(content: Content) -> some View {
        content
            .foregroundStyle(Color.gpInk)
            .padding(.horizontal, 14)
            .padding(.vertical, 12)
            .background(Color.gpSurface2, in: RoundedRectangle(cornerRadius: GP.radius))
            .overlay(
                RoundedRectangle(cornerRadius: GP.radius)
                    .strokeBorder(Color.gpBorder, lineWidth: 1))
    }
}

extension View {
    func ghostField() -> some View { modifier(GhostFieldStyle()) }

    /// Intitulé de section : petites capitales espacées, comme sur le web.
    func sectionLabel() -> some View {
        font(.caption.weight(.semibold))
            .textCase(.uppercase)
            .kerning(0.6)
            .foregroundStyle(Color.gpMuted)
    }
}

// ─── Structures d'écran ───

/// Un écran de la suite : le fond de nuit, et un contenu qui défile par-dessus.
struct GhostScreen<Contenu: View>: View {
    var espacement: CGFloat = 22
    @ViewBuilder var contenu: () -> Contenu

    var body: some View {
        ZStack {
            GhostBackground()
            ScrollView {
                VStack(alignment: .leading, spacing: espacement) {
                    contenu()
                }
                .padding(.horizontal, 20)
                .padding(.top, 8)
                .padding(.bottom, 40)
                .frame(maxWidth: 560, alignment: .leading)
                .frame(maxWidth: .infinity)
            }
            .scrollDismissesKeyboard(.interactively)
        }
    }
}

/// Une section : un intitulé en petites capitales, puis une carte qui réunit ses lignes.
struct GhostSection<Contenu: View>: View {
    // Clefs de traduction, pas des chaînes : ce qui s'écrit ici s'affiche, et doit donc
    // passer par le catalogue. Une `String` s'afficherait telle quelle, en français.
    var titre: LocalizedStringKey?
    var note: LocalizedStringKey?
    @ViewBuilder var contenu: () -> Contenu

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            if let titre { Text(titre).sectionLabel() }
            VStack(spacing: 0) { contenu() }
                .background(
                    Color.gpSurface.opacity(0.7),
                    in: RoundedRectangle(cornerRadius: GP.radiusCard)
                )
                .overlay(
                    RoundedRectangle(cornerRadius: GP.radiusCard)
                        .strokeBorder(Color.gpBorder, lineWidth: 1))
            if let note {
                Text(note)
                    .font(.caption)
                    .foregroundStyle(Color.gpMuted)
                    .padding(.horizontal, 2)
            }
        }
    }
}

/// Un filet de séparation entre deux lignes d'une même carte.
struct GhostDivider: View {
    var body: some View {
        Rectangle()
            .fill(Color.gpBorder)
            .frame(height: 1)
            .padding(.leading, 14)
    }
}

/// Une ligne de consultation : un intitulé, une valeur, et ce qu'on peut en faire.
struct GhostRow<Actions: View>: View {
    let intitule: LocalizedStringKey
    /// Le contenu du coffre, jamais traduit : un mot de passe reste ce qu'il est.
    let valeur: String
    var monospace = false
    var estSecret = false
    @ViewBuilder var actions: () -> Actions

    var body: some View {
        HStack(alignment: .firstTextBaseline, spacing: 12) {
            VStack(alignment: .leading, spacing: 3) {
                Text(intitule)
                    .font(.caption)
                    .foregroundStyle(Color.gpMuted)
                // `.enabled` et `.disabled` sont deux types distincts : un ternaire ne
                // les unifie pas. Le modificateur ne porte donc que sur ce qui se copie.
                Group {
                    if estSecret {
                        Text(verbatim: valeur)
                    } else {
                        Text(verbatim: valeur).textSelection(.enabled)
                    }
                }
                .font(monospace ? .system(.body, design: .monospaced) : .body)
                .foregroundStyle(Color.gpInk)
                .fixedSize(horizontal: false, vertical: true)
            }
            Spacer(minLength: 8)
            HStack(spacing: 14) { actions() }
                .foregroundStyle(Color.gpAccentText)
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 12)
    }
}

/// Bouton d'action discret d'une ligne : une icône, une zone de frappe correcte.
struct GhostIconButton: View {
    let systemImage: String
    var action: () -> Void

    var body: some View {
        Button(action: action) {
            Image(systemName: systemImage)
                .font(.system(size: 16, weight: .medium))
                .frame(width: 30, height: 30)
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .foregroundStyle(Color.gpAccentText)
    }
}

/// Ce que voit le sélecteur d'applications à la place du coffre : l'enseigne, rien d'autre.
///
/// iOS photographie l'écran quand l'application le quitte et garde cette vignette sur
/// disque. Un coffre déverrouillé s'y retrouverait en clair — noms de sites, identifiants —
/// visible d'un simple glissement, et lisible par qui inspecte le système de fichiers.
struct VoileDeConfidentialite: View {
    var body: some View {
        ZStack {
            GhostBackground()
            VStack(spacing: 12) {
                Image("LogoMark")
                    .resizable()
                    .scaledToFit()
                    .frame(width: 56, height: 56)
                Text("GhostPass")
                    .font(.system(.headline, design: .rounded, weight: .semibold))
                    .foregroundStyle(Color.gpMuted)
            }
        }
        .ignoresSafeArea()
        .transition(.opacity)
        .accessibilityIdentifier("view.privacyVeil")
    }
}

File: apps/android/app/src/main/kotlin/ch/stackops/ghostpass/theme/Theme.kt
package ch.stackops.ghostpass.theme

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.blur
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.ExperimentalComposeUiApi
import androidx.compose.ui.semantics.testTagsAsResourceId
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * Le système visuel néon de la suite, porté depuis `apps/ios/Ghostpass/Theme/Theme.swift`.
 *
 * La structure est commune à tous les produits : une base sombre dérivée de Dracula, des
 * surfaces de verre fumé, et un halo coloré diffusé depuis le haut. Ce qui les distingue
 * est leur teinte néon, celle de leur logo — pour GhostPass, le **bleu `#2E7DFF`**.
 *
 * Le néon n'est pas dans la couleur mais dans son rayonnement : deux halos superposés,
 * l'un serré et vif, l'autre large et discret. Voir [neon].
 *
 * **Ce fichier ne reprend pas seulement la palette, mais le langage visuel.** C'est la
 * règle §10 du brief, et elle vient d'un retour de Clara sur le portage Flutter de
 * GhostCal — « oh c'est moche comparé à l'app de ghostpass ». Reprendre les couleurs sans
 * les formes donne des composants Material simplement recolorés, et ça se voit
 * immédiatement : c'est le rayon des cartes, l'épaisseur du filet, la surface creusée des
 * champs et le halo des boutons qui font l'identité, pas le bleu.
 */

// ─── Base sombre commune à la suite (Dracula) ───

private val BaseSombre = Color(0xFF21222C)
private val BaseClair = Color(0xFFF5F8FC)
private val SurfaceSombre = Color(0xFF282A36)
private val SurfaceClair = Color(0xFFFFFFFF)
private val Surface2Sombre = Color(0xFF323445)
private val Surface2Clair = Color(0xFFEEF3F9)
private val EncreSombre = Color(0xFFF8F8F2)
private val EncreClair = Color(0xFF0F1B2D)
private val AttenueSombre = Color(0xFF8B9CC8)
private val AttenueClair = Color(0xFF5B6B82)

// ─── Teinte de marque : bleu néon ───

private val AccentSombre = Color(0xFF2E7DFF)
private val AccentClair = Color(0xFF1A4FCC)
private val AccentTexteSombre = Color(0xFF7FB2FF)
private val AccentTexteClair = Color(0xFF1A4FCC)

/**
 * La teinte du halo. **Fixe** : un néon garde sa couleur, c'est ce qui le fait lire comme
 * une source lumineuse plutôt que comme une ombre portée teintée.
 */
val Neon = Color(0xFF2E7DFF)

private val DangerSombre = Color(0xFFFF5555)
private val DangerClair = Color(0xFFD11F45)
private val SuccesSombre = Color(0xFF50FA7B)
private val SuccesClair = Color(0xFF15803D)

/** Les couleurs du produit, au-delà de ce que porte [MaterialTheme]. */
data class CouleursGhost(
    val base: Color,
    val surface: Color,
    val surface2: Color,
    val encre: Color,
    val attenue: Color,
    val accent: Color,
    val accentTexte: Color,
    val surAccent: Color,
    val bordure: Color,
    val bordureForte: Color,
    val danger: Color,
    val succes: Color,
    val sombre: Boolean,
)

val LocalCouleurs = staticCompositionLocalOf {
    couleursDe(sombre = true)
}

private fun couleursDe(sombre: Boolean) = CouleursGhost(
    base = if (sombre) BaseSombre else BaseClair,
    surface = if (sombre) SurfaceSombre else SurfaceClair,
    surface2 = if (sombre) Surface2Sombre else Surface2Clair,
    encre = if (sombre) EncreSombre else EncreClair,
    attenue = if (sombre) AttenueSombre else AttenueClair,
    accent = if (sombre) AccentSombre else AccentClair,
    accentTexte = if (sombre) AccentTexteSombre else AccentTexteClair,
    surAccent = Color.White,
    bordure = if (sombre) Color.White.copy(alpha = 0.08f) else Color(0xFF0F172A).copy(alpha = 0.10f),
    bordureForte = if (sombre) Color.White.copy(alpha = 0.16f) else Color(0xFF0F172A).copy(alpha = 0.16f),
    danger = if (sombre) DangerSombre else DangerClair,
    succes = if (sombre) SuccesSombre else SuccesClair,
    sombre = sombre,
)

/**
 * Mesures partagées, relevées dans `suite/docs/charte-mobile.md` §5.
 *
 * **Les jetons de `ghost-theme.css` font foi, pas l'implémentation d'iOS.** La charte le
 * dit d'expérience : sa première version avait été écrite en observant GhostPass iOS, qui
 * avait dérivé du web — cartes à 16 au lieu de 18 — et un document écrit en observant une
 * implémentation ne fait que graver sa dérive. Les valeurs ci-dessous viennent donc du CSS.
 *
 * Trois rayons distincts, et l'écart entre eux est délibéré : **un champ moins arrondi
 * qu'un bouton creuse la hiérarchie** entre ce qu'on remplit et ce sur quoi on appuie.
 * Les aplatir sur une seule valeur ferait disparaître cette lecture.
 */
object GP {
    /** `--radius-sm` : les champs de saisie. */
    val rayonChamp: Dp = 8.dp

    /**
     * `--radius` : les boutons.
     *
     * **Pas `--radius-pill`**, et c'est une exception assumée de la charte (§5, §7) : le web
     * arrondit ses boutons en pilule, iOS les garde rectangulaires parce que sa barre de
     * recherche appartient au système et ne peut pas devenir une pilule. Android reprend le
     * rectangle — non par imitation d'iOS, mais parce que le champ de recherche de Material
     * n'est pas une pilule non plus, et qu'un bouton en pilule posé à côté jurerait de la
     * même façon. Le motif est le même ; il a été revérifié plutôt que transposé.
     */
    val rayon: Dp = 12.dp

    /** `--radius-lg` : les cartes. */
    val rayonCarte: Dp = 18.dp

    val ecart: Dp = 12.dp
    val marge: Dp = 16.dp
    /** Marge intérieure d'une carte — absente du CSS, propre au mobile. */
    val margeCarte: Dp = 24.dp
    /** Écart entre deux champs. */
    val ecartChamps: Dp = 18.dp
    /** Écart d'un libellé à son champ. */
    val ecartLibelle: Dp = 7.dp
    /** Au-delà, sur tablette, les champs s'étirent et le formulaire perd sa forme. */
    val largeurMax: Dp = 420.dp
}

/**
 * **`testTagsAsResourceId` est posé ici, et c'est ce qui rend les `testTag` observables.**
 *
 * Sans ce drapeau, un `testTag` n'existe que pour les tests Compose ; il ne remonte dans
 * l'arbre d'accessibilité d'Android ni pour UiAutomator, ni pour `uiautomator dump`. Avec
 * lui, chaque `testTag` devient le `resource-id` du nœud — un champ qui **ne se prononce
 * pas**, et que `By.res` sait lire.
 *
 * Il est posé à la racine du thème plutôt qu'à chaque activité : les deux activités du
 * produit — l'écran principal et celui du remplissage — passent par ici, et l'oublier dans
 * l'une des deux rendrait ses témoins aveugles **sans rien afficher d'anormal**. C'est
 * exactement la classe de panne silencieuse qu'on cherche à éviter : un témoin qui ne
 * trouve plus rien accuse le produit.
 *
 * ## Vérifié à l'écran, et le premier contrôle mentait
 *
 * `uiautomator dump` montre bien les `resource-id` — le `FLAG_SECURE` du produit noircit les
 * captures d'image, pas l'arbre d'accessibilité. Le premier dépouillement n'en a pourtant
 * montré aucun, et la conclusion « le drapeau ne se propage pas aux descendants » était
 * fausse : **l'APK installé était périmé**. `:app:assembleDebugAndroidTest` compile bien le
 * paquet de test, mais ne reconstruit pas `app-debug.apk` ; l'installer après lui revient à
 * mesurer le code d'avant.
 *
 * Ce n'est pas une anecdote de construction, c'est l'instrument qui n'a pas su rougir : il a
 * rendu un arbre parfaitement bien formé, d'une application parfaitement lancée, et rien
 * dans sa sortie ne disait de quelle version. Reconstruire explicitement et regarder
 * l'horodatage de l'APK avant de l'installer coûte une ligne.
 */
@OptIn(ExperimentalComposeUiApi::class)
@Composable
fun ThemeGhostPass(
    sombre: Boolean = isSystemInDarkTheme(),
    contenu: @Composable () -> Unit,
) {
    val couleurs = couleursDe(sombre)
    // On suit le réglage du système plutôt que d'imposer l'un des deux thèmes, comme sur le
    // web et sur iOS.
    val schema = if (sombre) {
        darkColorScheme(
            primary = couleurs.accent,
            onPrimary = couleurs.surAccent,
            background = couleurs.base,
            surface = couleurs.surface,
            onSurface = couleurs.encre,
            error = couleurs.danger,
        )
    } else {
        lightColorScheme(
            primary = couleurs.accent,
            onPrimary = couleurs.surAccent,
            background = couleurs.base,
            surface = couleurs.surface,
            onSurface = couleurs.encre,
            error = couleurs.danger,
        )
    }
    CompositionLocalProvider(LocalCouleurs provides couleurs) {
        MaterialTheme(colorScheme = schema) {
            Box(Modifier.semantics { testTagsAsResourceId = true }) { contenu() }
        }
    }
}

/**
 * Le fond de l'application : nuit profonde et halo néon diffusé depuis le haut — la lueur
 * du fantôme, qui donne sa profondeur à l'ensemble sans rien coûter en lisibilité.
 *
 * Le dégradé du fond est celui de la suite : quatre arrêts, teintés de la couleur de
 * marque, qui empêchent le noir d'être plat. Le halo est un dégradé radial centré en haut,
 * flouté — c'est lui qui distingue cet écran d'un fond uni sombre.
 */
@Composable
fun FondGhost(modifier: Modifier = Modifier) {
    val couleurs = LocalCouleurs.current
    val fond = if (couleurs.sombre) {
        Brush.verticalGradient(
            listOf(
                Color(0xFF080D18), Color(0xFF0A1220),
                Color(0xFF080F1A), Color(0xFF090A10),
            ),
        )
    } else {
        Brush.verticalGradient(listOf(couleurs.base, couleurs.base))
    }
    Box(modifier.fillMaxSize().background(fond)) {
        Box(
            Modifier
                .fillMaxWidth()
                .padding(bottom = 0.dp)
                .background(
                    Brush.radialGradient(
                        colors = listOf(
                            Neon.copy(alpha = if (couleurs.sombre) 0.26f else 0.12f),
                            Color.Transparent,
                        ),
                        center = Offset(0.5f, 0f),
                        radius = 900f,
                    ),
                )
                .blur(40.dp),
        )
    }
}

/**
 * Carte de verre fumé : la surface translucide, bordée d'un filet clair.
 *
 * L'opacité de 0,65 n'est pas un détail : c'est ce qui laisse transparaître le halo du
 * fond et fait lire la carte comme posée *sur* quelque chose. Une surface opaque donnerait
 * une boîte, et l'écran d'entrée ressemblerait à n'importe quel formulaire.
 *
 * **Écart avec la charte §4, assumé et à inscrire au tableau §7 : le flou d'arrière-plan
 * manque.** SwiftUI le donne (`.ultraThinMaterial`) ; Compose n'a aucun équivalent. Son
 * `Modifier.blur` floute le composant *lui-même*, pas ce qui est derrière — l'appliquer ici
 * rendrait les champs illisibles au lieu d'adoucir le fond. Un vrai flou d'arrière-plan
 * demande `RenderEffect`, disponible seulement à partir de l'API 31, alors que le `minSdk`
 * du produit est 24 ; le faire dépendre de la version donnerait deux apparences selon
 * l'appareil, ce qui est pire qu'une seule apparence assumée.
 *
 * Ce que la charte redoute — « sans lui la carte n'est qu'un rectangle gris » — est
 * partiellement évité ici parce que le fond n'est pas uni : c'est un dégradé portant un
 * halo radial, et la translucidité en laisse voir la variation. La carte lit donc comme une
 * surface posée sur une lueur, à défaut de lire comme du verre.
 */
@Composable
fun Modifier.carteDeVerre(marge: Dp = GP.margeCarte): Modifier {
    val couleurs = LocalCouleurs.current
    val forme = RoundedCornerShape(GP.rayonCarte)
    return this
        .background(couleurs.surface.copy(alpha = 0.65f), forme)
        .border(1.dp, couleurs.bordure, forme)
        .padding(marge)
}

/**
 * Champ de saisie : surface creusée, bordure discrète, au rayon des **champs** (8).
 *
 * Moins arrondi que les boutons (12), délibérément : c'est ce qui distingue à l'œil ce
 * qu'on remplit de ce sur quoi on appuie.
 */
@Composable
fun Modifier.champGhost(): Modifier {
    val couleurs = LocalCouleurs.current
    val forme = RoundedCornerShape(GP.rayonChamp)
    return this
        .background(couleurs.surface2, forme)
        .border(1.dp, couleurs.bordure, forme)
}

/** La famille de caractères arrondie de l'enseigne, comme le `design: .rounded` d'iOS. */
val PoliceArrondie = FontFamily.SansSerif

/**
 * Fait rayonner un élément dans la teinte de la marque.
 *
 * Transposition du `--brand-glow` partagé — `drop-shadow(0 0 8px …0.6)` puis
 * `drop-shadow(0 0 20px …0.35)`. **Les deux comptent** : un seul halo fait une tache molle,
 * tandis que la superposition d'un noyau net et d'une aura étalée est ce qui donne
 * l'impression d'une source de lumière. Le rayon est réduit en thème clair, où un néon sur
 * fond blanc devient une bavure.
 *
 * Dessiné à la main plutôt que confié à `Modifier.shadow` : les couleurs d'ombre de Compose
 * ne sont honorées qu'à partir de l'API 28, et le `minSdk` du produit est 24. Sur un
 * appareil ancien, `shadow` aurait rendu une ombre grise — c'est-à-dire de la saleté sous
 * un bouton, là où on voulait de la lumière autour. Personne ne l'aurait vu depuis un
 * émulateur récent.
 */
@Composable
fun Modifier.neon(force: Float = 1f, rayon: Dp = GP.rayon): Modifier {
    val couleurs = LocalCouleurs.current
    val echelle = if (couleurs.sombre) force else force * 0.45f
    if (echelle <= 0f) return this
    return this.drawBehind {
        val forme = androidx.compose.ui.geometry.Size(size.width, size.height)
        // L'aura large et discrète, puis le noyau serré et vif, dans cet ordre : le second
        // se pose sur le premier.
        for ((etalement, opacite) in listOf(20.dp to 0.35f, 8.dp to 0.60f)) {
            val marge = etalement.toPx() * echelle
            drawRoundRect(
                brush = Brush.radialGradient(
                    colors = listOf(
                        Neon.copy(alpha = opacite * echelle),
                        Color.Transparent,
                    ),
                    center = androidx.compose.ui.geometry.Offset(
                        forme.width / 2f, forme.height / 2f),
                    radius = (maxOf(forme.width, forme.height) / 2f) + marge,
                ),
                topLeft = androidx.compose.ui.geometry.Offset(-marge, -marge),
                size = androidx.compose.ui.geometry.Size(
                    forme.width + marge * 2, forme.height + marge * 2),
                cornerRadius = androidx.compose.ui.geometry.CornerRadius(
                    (rayon.toPx() + marge)),
            )
        }
    }
}
